server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2545,33 +2545,37 @@ public void addShardFailureCallback(Consumer<ShardFailure> onShardFailure) {
      * indexing operation, so we can flush the index.
      */
     public void flushOnIdle(long inactiveTimeNS) {
-        Engine engineOrNull = getEngineOrNull();
-        if (engineOrNull != null && System.nanoTime() - engineOrNull.getLastWriteNanos() >= inactiveTimeNS) {
-            boolean wasActive = active.getAndSet(false);
-            if (wasActive) {
-                logger.debug("flushing shard on inactive");
-                threadPool.executor(ThreadPool.Names.FLUSH)
-                    .execute(() -> flush(new FlushRequest().waitIfOngoing(false).force(false), new ActionListener<>() {
-                        @Override
-                        public void onResponse(Engine.FlushResult flushResult) {
-                            if (flushResult.skippedDueToCollision()) {
-                                // In case an ongoing flush was detected, revert active flag so that a next flushOnIdle request
-                                // will retry (#87888)
-                                active.set(true);
+        tryWithEngineOrNull(engineOrNull -> {
+            if (engineOrNull != null && System.nanoTime() - engineOrNull.getLastWriteNanos() >= inactiveTimeNS) {
+                boolean wasActive = active.getAndSet(false);
+                if (wasActive) {
+                    logger.debug("flushing shard on inactive");
+                    threadPool.executor(ThreadPool.Names.FLUSH)
```

Comment:
Can I pick your brain here: there's a rare edge case that tryWithEngine gets the engine, comes here, but afterwards the flush thread itself tries to get the engine while it might be again being reset, getting stuck. And sometimes resetting the engine might need to flush. So it could deadlock.

We could solve this either by passing the tryWithEngine's engine down to to here, so it'd be the older engine. Or, as I had attempted before, to make the flush with waitIfOngoing=false use tryWithEngine.

Or, we just don't deal with this PR at all.

Would appreciate your feedback!